modules/data.atmosphere/R/downscale_spline_to_hourly.R

```diff
@@ -0,0 +1,56 @@
+##' @title Downscale spline to hourly
+##' @return A dataframe of downscaled state variables
+##'
+##' @param df, dataframe of data to be downscales 
+##' @param VarNamesStates, names of vars that are state variables
+##' @export
+##' 
+##' @author Laura Puckett
+##' 
+##' 
+
+downscale_spline_to_hourly <- function(df,VarNamesStates){
+  # --------------------------------------
+  # purpose: interpolates debiased forecasts from 6-hourly to hourly
+  # Creator: Laura Puckett, December 16 2018
+  # --------------------------------------
+  # @param: df, a dataframe of debiased 6-hourly forecasts
+
+  interpolate <- function(jday, var){
+    result <- splinefun(jday, var, method = "monoH.FC")
+    return(result(seq(min(as.numeric(jday)), max(as.numeric(jday)), 1/24)))
+  }
+  
+
+  
+  t0 = min(df$timestamp)
+  df <- df %>%
+    dplyr::mutate(days_since_t0 = difftime(.$timestamp, t0, units = "days"))
+  
+  if("dscale.member" %in% colnames(df)){
+    by.ens <- df %>% 
+      dplyr::group_by(NOAA.member, dscale.member)
+  }else{
+    by.ens <- df %>% 
+      dplyr::group_by(NOAA.member)
+    }
+  
+  interp.df.days <- by.ens %>% dplyr::do(days = seq(min(df$days_since_t0), as.numeric(max(df$days_since_t0)), 1/24))
+  interp.df <- interp.df.days
+  
+  for(Var in 1:length(VarNamesStates)){
+    assign(paste0("interp.df.",VarNamesStates[Var]), dplyr::do(by.ens, var = interpolate(.$days_since_t0,unlist(.[,VarNamesStates[Var]]))) %>% plyr::rename(c("var" = VarNamesStates[Var])))
+    if("dscale.member" %in% colnames(df)){
+        interp.df <- dplyr::inner_join(interp.df, get(paste0("interp.df.",VarNamesStates[Var])), by = c("NOAA.member", "dscale.member"))
+    }else{
+      interp.df <- dplyr::inner_join(interp.df, get(paste0("interp.df.",VarNamesStates[Var])), by = c("NOAA.member"))
+    } 
```

Comment:
Using `assign` and `get` like this makes me a bit nervous. In most cases (including here), I think a named list can accomplish the same thing; i.e.

```r
var_list <- list()
for (Var in seq_along(VarNamesStates) {
  var_list_name <- paste0("inerp.df.", VarNamesStates[[Var]])
  var_list[[var_list_name]] <- dplyr::do(...)
  ...
  interp.df <- dplyr::inner_join(interp.df, var_list[[var_list_name]], by = ...)
```

But, if it works consistently, we can leave it for now. 

I _would_ use `dplyr::rename` instead of `plyr::rename` though -- `plyr` is a largely deprecated package that we should gradually remove from our code base in favor of more modern, better-maintained alternatives. You should be able to replace `plyr::rename(c("var" = VarNamesStates[Var]))` with `dplyr::rename(!!VarNamesStates[Var] := "var")`.

And actually, with modern `dplyr`, you can get away with skipping the `rename` step altogether by writing this as:

```r
dplyr::do(by.ens, !!VarNamesStates[[Var]] := interpolate(...))
```

(In a nutshell, in `dplyr` and related functions, `!!x := y` basically means "evaluate y and assign the result to a variable whose name is the value of x". If you're curious, this is called "Quasiquotation", and you can read more about it in the associated chapter of [Advanced R](https://adv-r.hadley.nz/quasiquotation.html#unquoting).).